plutus-core/testlib/PlutusIR/Generators/QuickCheck/ShrinkTerms.hs

```diff
@@ -275,22 +269,34 @@ shrinkTypedTerm tyctx0 ctx0 (ty0, tm0) = concat
           | TyFun _ _ b <- [ty]
           ]
 
-        Apply _ fun arg | Right argTy <- inferTypeInContext tyctx ctx arg ->
-          -- Drop substerms
-          [(argTy, arg), (TyFun () argTy ty, fun)]
+        Apply _ fun arg -> case inferTypeInContext tyctx ctx arg of
+          Right argTy ->
+            -- Drop substerms
+            [ -- Prefer to keep the same type, but...
+              (ty, fixupTerm tyctx ctx tyctx ctx ty arg)
+            , (ty, fixupTerm tyctx ctx tyctx ctx ty fun)
+            , -- ... support type-changing shrinking as well.
+              (argTy, arg)
+            , (TyFun () argTy ty, fun)
+            ]
+          Left err -> error $ displayPlcCondensedErrorClassic err
 
         TyAbs _ x _ body ->
           [ fixupTerm_ (Map.insert x k tyctx) ctx tyctx ctx tyInner' body
           | TyForall _ y k tyInner <- [ty]
           , let tyInner' = typeSubstClosedType y (minimalType k) tyInner
           ]
 
-        -- Builtins can shrink to unit. More fine-grained shrinking is in `structural` below.
-        Constant _ val -> do
-            val'@(Some (ValueOf uni _)) <- shrink val
-            pure (mkTyBuiltinOf () uni, Constant () val')
-
-        _ -> []
+        -- TODO: allow non-structural shrinking for some of these.
+        Var{} -> []
```

Comment:
I expanded the underscore just to make it clear what kind of shrinks we're missing.